frontend/src/metabase/common/components/FilterPicker/FilterPicker.tsx

```diff
@@ -77,26 +74,25 @@ export function FilterPicker({
     [filterIndex],
   );
 
-  const handleExpressionChange = useCallback(
-    (name: string, expression: LegacyExpressionClause) => {
-      if (Array.isArray(expression) && isLegacyExpression(expression)) {
-        const baseFilter =
-          legacyFilter || new LegacyFilter([], null, legacyQuery);
-        const nextFilter = baseFilter.set(expression);
-        onSelectLegacy(nextFilter);
-        onClose?.();
-      }
+  const handleClauseChange = useCallback(
+    (_name: string, clause: Lib.CustomExpressionClause) => {
+      onSelect(
+        clause as Exclude<Lib.CustomExpressionClause, Lib.AggregationClause>,
```

Comment:
+1 types aren't correct

Reply:
I managed to avoid the casting by making `ExpressionWidget` generic: b9a82b65a37c9327dbbfff0937da5b718e7d3f65

I don't think it's possible to avoid the casts [otherwise](https://github.com/metabase/metabase/pull/36396#discussion_r1415496134).